### PR TITLE
Add $facet tests

### PR DESCRIPTION
**Pull request by PatersonProjects**
This pr adds tests for the $facet stage operator

Ref: Issue #47, #485

### CODE REVIEW

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        expected=test_case.expected,
+        error_code=test_case.error_code,
+        msg=test_case.msg,
+        ignore_doc_order=test_case.ignore_doc_order,
+        ignore_order_in=test_case.ignore_order_in,
+    )
+
+
+@pytest.mark.aggregate
+def test_facet_subpipeline_lookup_basic(collection):
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:572]: The `$lookup` test only covers the equi-join form (`localField`/`foreignField`). Missing: `$lookup` with pipeline form inside `$facet` — this tests nested sub-pipelines (`$facet` → `$lookup` → `pipeline` → [stages]). Consider adding:

1. Success tests with pipeline-form `$lookup` containing multiple stages (e.g., `$match`, `$project`, `$sort`, `$group`, `$unwind`, `$addFields`, etc.)
2. Error tests with all `$facet`-forbidden stages (`$out`, `$merge`, `$collStats`, `$indexStats`, `$planCacheStats`, `$geoNear`, `$facet`) inside the `$lookup` pipeline to verify they're still rejected when nested inside `$facet`

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:1]: Many success/behavior tests (across the whole PR) use a single sub-pipeline. `$facet` is designed for multiple sub-pipelines on the same input — single-sub-pipeline is a special/minimal case. Consider adding a second sub-pipeline (e.g., `"first": [{"$limit": 1}]`) to the success cases in `test_facet_subpipeline_stages.py`, `test_stages_position_facet.py`, and `test_facet_collation.py` to make them more realistic.

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+            "byMissing": Eq([{"_id": None, "count": 4}]),
+            "autoMissing": [Len(1), Eq([{"_id": {"min": None, "max": None}, "count": 4}])],
+        },
+        msg="Sub-pipelines on a non-existent field should not crash and return one doc each",
+    ),
+]
+
+
+@pytest.mark.aggregate
+@pytest.mark.parametrize("test_case", pytest_params(FACET_SUBPIPELINE_STAGE_TESTS))
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:550]: Per the container-feature rule (one test per sub-feature), these stages are valid inside `$facet` but not tested: `$redact`, `$sample`, `$unionWith`, `$setWindowFields`. Each should have one smoke case confirming it works inside a `$facet` sub-pipeline.

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_argument_validation.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+from documentdb_tests.framework.executor import execute_command
+from documentdb_tests.framework.parametrize import pytest_params
+
+DOCS = [{"_id": 1, "cat": "A"}, {"_id": 2, "cat": "B"}]
+
+LONG_NAME = "f" * 300
+
+# Property [Valid Edge Cases]: empty sub-pipelines, many sub-pipelines, and
+# unusual-but-valid output field names are accepted.
+FACET_ARGUMENT_SUCCESS_TESTS: list[StageTestCase] = [
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_argument_validation.py:28]: Missing field name edge cases that should be verified (as either accepted or rejected): spaces (`"my field"`), hyphens (`"my-field"`), underscores (`"my_field"`), numeric-only (`"123"`), at-sign (`"@field"`), null byte (`"a\x00b"`). These should be tested to document the actual behavior — whether they succeed or error.

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_core_semantics.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    ),
+    StageTestCase(
+        id="skip_beyond_count_empty",
+        docs=DOCS,
+        pipeline=[{"$facet": {"a": [{"$skip": 100}]}}],
+        expected=[{"a": []}],
+        msg="$facet should yield an empty array when $skip exceeds the document count",
+    ),
+    StageTestCase(
+        id="allowDiskUse_true_operates_normally",
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_core_semantics.py:128]: Missing the `allowDiskUse: false` case. Both `true` and `false` should be tested to confirm `$facet` operates normally regardless of the flag value.

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_independence.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+DOCS = [
+    {"_id": 1, "cat": "A", "v": 10},
+    {"_id": 2, "cat": "A", "v": 20},
+    {"_id": 3, "cat": "B", "v": 30},
+]
+
+# Property [Sub-Pipeline Independence]: a $limit, $match, or $addFields in one
+# sub-pipeline does not affect the input seen by another; a runtime error in any
+# sub-pipeline fails the whole stage.
+FACET_INDEPENDENCE_TESTS: list[StageTestCase] = [
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_independence.py:31]: Please add two tests:
1. `$sample` before `$facet` — two sub-pipelines that each return all documents and verify they see exactly the same data
2. `$addFields` with `$rand` before `$facet` — two sub-pipelines that each return all documents and verify they see exactly the same `$rand` values

**Review by fanyangv — CHANGES_REQUESTED** [review]

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_independence.py (hunk 5 above)
Comment (PatersonProjects) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_independence.py:31]: Added!

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_core_semantics.py (hunk 4 above)
Comment (PatersonProjects) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_core_semantics.py:128]: Added an allowDiskUse: false case

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_argument_validation.py (hunk 3 above)
Comment (PatersonProjects) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_argument_validation.py:28]: Added spaces, hyphens, underscores, atsign, and numeric case, all succeed. Null byte gets caught by pymongo, causing unreliable test cases.

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py (hunk 2 above)
Comment (PatersonProjects) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:550]: Added tests

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py
Comment (PatersonProjects) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:1]: Added second sub-pipeline using count to most single-pipeline success cases

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py (hunk 1 above)
Comment (PatersonProjects) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:572]: Added!

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_independence.py (hunk 5 above)
Comment (fanyangv) [reply documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_independence.py:31]: Thanks! `$sample: {size: 10}` on 3 documents returns all docs — making `$sample` effectively a no-op here. To properly verify both sub-pipelines see the same sampled subset, please use more documents than the sample size (e.g., 20 docs with `$sample: {size: 5}`).

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py [hunk 6]
```diff
@@ -0,0 +1,856 @@
...
+DOCS = [
+    {"_id": 1, "cat": "A", "price": 10, "tags": ["x", "y"]},
+    {"_id": 2, "cat": "A", "price": 20, "tags": ["y"]},
+    {"_id": 3, "cat": "B", "price": 30, "tags": ["z"]},
+    {"_id": 4, "cat": "C", "price": 40, "tags": []},
+]
+
+# Property [Sub-Pipeline Stage Support]: common aggregation stages and
+# use-case combinations produce the expected output arrays inside $facet.
+FACET_SUBPIPELINE_STAGE_TESTS: list[StageTestCase] = [
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_subpipeline_stages.py:34]: Missing supported stages inside `$facet`: `$densify`, `$fill`, `$documents` should each have one smoke case. Also missing forbidden stages (non-Atlas): `$currentOp` and `$changeStream` should be tested as rejected inside `$facet`.

**Review by fanyangv — CHANGES_REQUESTED** [review]
Thanks for the great test coverage! Last changes requested are around: FACET_SUBPIPELINE_STAGE_TESTS; $sample before $facet case.

File: documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_errors.py [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+        docs=CORE_DOCS,
+        pipeline=[{"$facet": {"a": [{"$limit": 0}]}}],
+        error_code=LIMIT_NOT_POSITIVE_ERROR,
+        msg="$facet should reject $limit 0 inside a sub-pipeline",
+    ),
+]
+
+# Property [Forbidden Sub-Pipeline Stages]: stages that cannot run inside a
+# $facet sub-pipeline are rejected with 40600, independent of documents.
+FACET_FORBIDDEN_STAGE_TESTS: list[StageTestCase] = [
```
Comment (fanyangv) [inline documentdb_tests/compatibility/tests/core/operator/stages/facet/test_facet_errors.py:257]: Could you also add $listLocalSessions to the forbidden stage tests? This is a stage that should be rejected inside $facet.

**Review by fanyangv — CHANGES_REQUESTED** [review]

**Review by fanyangv — APPROVED** [review]

**Review by eerxuan — APPROVED** [review]
